xds/internal/xdsclient/authority_test.go

```diff
@@ -0,0 +1,344 @@
+	cancelResource := a.watchResource(listenerResourceType, resourceName, watcher)
+	defer cancelResource()
+
+	// Waiting for mgmt server to recv request.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before mgmt server got the request.")
+	case <-onStreamRequestDoneCh:
+	}
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateRequested); err != nil {
+		t.Fatal(err)
+	}
+
+	mgmtServer.Stop()
+
+	// Waiting for watcher to receive the update.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before verifying error propagation.")
+	case err := <-watcher.errorCh:
+		if err == nil {
+			t.Fatal("got err == nil. Want: err from stream connection reset by peer")
+		}
+	}
+	if err := compareWStateAndWTimer(a, listenerResourceType, resourceName, watchStateStarted); err != nil {
+		t.Fatal(err)
+	}
+}
+
+// This tests the case where ADS stream breaks after a successfully receiving a
+// message on the stream. In this case, we do not want the error to be propagated
+// to the watchers. verifies that in the case where the ADS stream breaks after successfully
+// receiving a message on the stream. In this case we want to ignore propagating
+// connection error to the watcher. But since the mgmt server stops in this test,
+// the watcher would be updated with a connection error for the subsequent attempt
+// to create an ADS stream.
+func (s) TestWatchResourceTimerCanRestartOnIgnoredADSRecvError(t *testing.T) {
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTestTimeout)
+	defer cancel()
+
+	onStreamRequestDoneCh := make(chan struct{})
+	mgmtServerOpts := mgmtServerOptWithOnRequestDoneCh(onStreamRequestDoneCh)
+	a, mgmtServer, nodeID := setupAuthorityWithMgmtServer(ctx, t, mgmtServerOpts, defaultTestWatchExpiryTimeout)
+	defer a.close()
+
+	resourceNameA := "xdsclient-test-lds-resourceA"
+	watcherA := newTestResourceWatcher()
+
+	cancelWatchResourceA := a.watchResource(listenerResourceType, resourceNameA, watcherA)
+	defer cancelWatchResourceA()
+
+	// Wait for mgmt server to recv request.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before mgmt server got the request.")
+	case <-onStreamRequestDoneCh:
+	}
+
+	resources := e2e.UpdateOptions{
+		NodeID:         nodeID,
+		Listeners:      []*v3listenerpb.Listener{e2e.DefaultClientListener(resourceNameA, "new-rds-resource")},
+		SkipValidation: true,
+	}
+
+	if err := mgmtServer.Update(ctx, resources); err != nil {
+		t.Fatalf("Failed to update management server with resources: %v, err: %v", resources, err)
+	}
+
+	// Verifying update for resource A is received.
+	select {
+	case <-ctx.Done():
+		t.Fatal("Test timed out before watcher received the update.")
+	case <-watcherA.updateCh:
+	}
+
+	resourceNameB := "xdsclient-test-lds-resourceB"
+	watcherB := newTestResourceWatcher()
+	cancelWatchResourceB := a.watchResource(listenerResourceType, resourceNameB, watcherB)
+	defer cancelWatchResourceB()
+
```

Comment:
With a short timeout, verify that the update callback for watcherB is not invoked here.